File: app-server/src/test/java/com/github/arhor/simple/expense/tracker/web/api/UserControllerTest.java
package com.github.arhor.simple.expense.tracker.web.api;

import lombok.val;

import org.junit.jupiter.api.Test;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.security.test.context.support.WithMockUser;

import com.github.arhor.simple.expense.tracker.model.UserResponseDTO;
import com.github.arhor.simple.expense.tracker.service.UserService;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.argThat;
import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.then;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@WebMvcTest(UserController.class)
class UserControllerTest extends BaseControllerTest {

    @MockBean
    private UserService userService;

    @Test
    void should_return_status_201_user_info_and_location_header_creating_new_user() throws Exception {
        // given
        val username = "username";
        val password = "Password1";
        val currency = "USD";

        val response = new UserResponseDTO();
        response.setId(1L);
        response.setUsername(username);
        response.setCurrency(currency);

        given(userService.createNewUser(any()))
            .willReturn(response);

        // when
        val result = http.perform(
            post("/api/users")
                .contentType("application/json")
                .content("""
                    {
                        "username": "%s",
                        "password": "%s",
                        "currency": "%s"
                    }
                    """.formatted(username, password, currency))
        );

        // then
        then(userService)
            .should()
            .createNewUser(
                argThat(request -> {
                    return username.equals(request.getUsername())
                        && password.equals(request.getPassword())
                        && currency.equals(request.getCurrency());
                })
            );

        result
            .andExpect(status().isCreated())
            .andExpect(jsonPath("$.id").value(response.getId()))
            .andExpect(jsonPath("$.username").value(response.getUsername()))
            .andExpect(jsonPath("$.currency").value(response.getCurrency()));
    }

    @Test
    @WithMockUser
    void should_return_status_200_and_expected_info_for_authenticated_user() throws Exception {
        // given
        val response = new UserResponseDTO();
        response.setId(1L);
        response.setUsername("user");
        response.setCurrency("USD");

        given(userService.determineUser(any()))
            .willReturn(response);

        // when
        val result = http.perform(get("/api/users?current"));

        // then
        then(userService)
            .should()
            .determineUser(argThat(this::authenticatedUser));

        result
            .andExpect(status().isOk())
            .andExpect(jsonPath("$.id").value(response.getId()))
            .andExpect(jsonPath("$.username").value(response.getUsername()))
            .andExpect(jsonPath("$.currency").value(response.getCurrency()));
    }
}

File: apps/server/src/test/kotlin/com/github/arhor/simple/expense/tracker/api/controller/UserControllerTest.kt
package com.github.arhor.simple.expense.tracker.api.controller

import com.github.arhor.simple.expense.tracker.model.UserRequestDTO
import com.github.arhor.simple.expense.tracker.model.UserResponseDTO
import com.github.arhor.simple.expense.tracker.service.UserService
import io.mockk.every
import io.mockk.slot
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.from
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.http.MediaType
import org.springframework.security.test.context.support.TestExecutionEvent
import org.springframework.security.test.context.support.WithUserDetails
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.post

@WebMvcTest(UserController::class)
internal class UserControllerTest : ControllerTestSupport() {

    @Autowired
    private lateinit var userService: UserService

    @Test
    fun `should return status 201, user info and location header creating new user`() {
        // given
        val user = slot<UserRequestDTO>()
        val expectedId = TestUser.id
        val expectedUsername = TestUser.username
        val expectedPassword = TestUser.password
        val expectedCurrency = TestUser.currency

        every { userService.createNewUser(request = any()) } returns UserResponseDTO(
            expectedId,
            expectedUsername,
            expectedCurrency
        )

        // when
        val awaitResult = http.post(appProps.apiUrlPath("/users")) {
            contentType = MediaType.APPLICATION_JSON
            content = """
                {
                    "username": "$expectedUsername",
                    "password": "$expectedPassword",
                    "currency": "$expectedCurrency"
                }
            """.trimIndent()
        }

        // then
        verify(exactly = 1) { userService.createNewUser(request = capture(user)) }

        assertThat(user.captured)
            .returns(expectedUsername, from { it.username })
            .returns(expectedPassword, from { it.password })
            .returns(expectedCurrency, from { it.currency })

        awaitResult.andExpect {
            status { isCreated() }
            jsonPath("$.id") { value(expectedId) }
            jsonPath("$.username") { value(expectedUsername) }
            jsonPath("$.currency") { value(expectedCurrency) }
        }
    }

    @Test
    @WithUserDetails(setupBefore = TestExecutionEvent.TEST_EXECUTION)
    fun `should return status 200 and expected info for authenticated user`() {
        // given
        val expectedId = TestUser.id
        val expectedUsername = TestUser.username
        val expectedCurrency = TestUser.currency

        // when
        val result = http.get(appProps.apiUrlPath("/users/current"))

        // then
        result.andExpect {
            status { isOk() }
            jsonPath("$.id") { value(expectedId) }
            jsonPath("$.username") { value(expectedUsername) }
            jsonPath("$.currency") { value(expectedCurrency) }
        }
    }
}
